### ISSUE TITLE
Tests should not fiddle with ~/.cookiecutterrc

### ISSUE THREAD

**Issue by vincentbernat** [cookiecutter/cookiecutter#517]
Hey!

Tests are directly using ~/.cookiecutterrc. They fail with mine. I see that there is code to move it out of the way, then move it back. It doesn't seem to work as expected. I am also a bit uncomfortable with tests that move files in my home. It would be better to set `HOME` environment variable to some temp directory instead, wouldn't it?

**Comment by pydanny** [comment]
:+1: to this as a problem.

**Comment by hackebrot** [comment]
I think all of the things we do in the ``clean_system`` fixture should not operate on the users' system.

:+1: for having an isolated test env!

**Comment by bionikspoon** [comment]
ref #623

This is a copy and paste from the pull request.  I wrote up a summary of changes.

## Summary of changes

Use temp directories instead of modifying the file system. 

The tests make use of pytest's `tmpdir` fixture.  Each test works in a unique folder instead of CWD.

### Modifications to the main `cookiecutter` module

- `cookiecutter.config`: replaced `USER_CONFIG_PATH` and `DEFAULT_CONFIG` with function call to 'lazy load' the values

  Reasoning: part of the fix is setting the `HOME` env variable.  However, these static variables are being called in too many places before the earliest test setup function can run.  

  It could've been monkeypatched I guess, but it would've had to been patched several places--everywhere it's being used, and this seems shady.

### Changes to the tests

- New global setup fixture that sets `HOME` env var and changes CWD for the test to the unique test folder.

- Improved cleanup routine, none.
  
  There was a ton of teardown/cleanup fixtures,  I deleted all of that.  With this setup you don't need it.   

  The concept: Each test gets a unique folder for that test.  Each time you run the tests, each of those unique folders are grouped in a folder for the run.  Pytest automatically preserves the trailing 5 test runs.

  tldr; pytest does the cleanup.

- absolute references.
  The problems created by using `tmpdir` and changing CWD:  references to the "tests" folder need to be absolute.   I created a helper function in a new file `tests.utils` called `dir_tests` to make this easier. It's just `os.join` partial.   And I basically did a search and replace  for references to the 'tests' folder.

  _Note: it was originally called `tests_dir`, which is way more conventional, but pytest was testing it._

#### Other changes to the tests

- removed a few gratuitous fixtures. Someone was going fixture happy :smile: 

  For example, there was this one that would return a string: "some_filename"  And another that would use that fixture to return "some_filename.json".  I think there was a 3rd one that used it, I forget now.

- In `tests.skipifmarkers`, I added `skipif_no_hg`.  I don't have hg, but I wanted passing tests; hope that's alright.

### Tests that are still failing

I need some help on these

- [`test_cli.test_run_cookiecutter_on_overwrite_if_exists_and_replay`](https://github.com/bionikspoon/cookiecutter/blob/fix-tests-fixtures/tests/test_cli.py#L124)

  `test_cli` was the hardest thing the get working.  `click` creates its own isolated environment.  I got all of them working but this one.  Between not knowing the code base very well and the error message, I have no idea where to start.

  ```pytb
>       assert result.exit_code == 0
E       assert -1 == 0
E        +  where -1 = <Result IOError(2, 'No such file or directory')>.exit_code
```


- [test_hooks](https://github.com/bionikspoon/cookiecutter/blob/fix-tests-fixtures/tests/test_hooks.py)

  This whole file hieroglyphics to me.  I tried looking at the documentation for [pytest's xunit setup](https://pytest.org/latest/xunit_setup.html) and got nowhere.  I'm wondering if you even need this test.   It looks strikingly similar to [`test_generate_hooks`](https://github.com/bionikspoon/cookiecutter/blob/fix-tests-fixtures/tests/test_generate_hooks.py).  If you do need it, does it have to use that class setup?

### Conclusion

It's not complete.  There's a few more tests that need fixed.

I know it's a bit invasive.  I thought it would be few lines here and there to fix, but it turned to be much more than that.

It does isolate the file system.  It does do what it's supposed to do.
